x/emissions/keeper/actor_penalties.go

```diff
@@ -0,0 +1,141 @@
+package keeper
+
+import (
+	alloraMath "github.com/allora-network/allora-chain/math"
+	"github.com/allora-network/allora-chain/x/emissions/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// MayPenaliseInferer penalises an inferer for missing previous epochs. It saves and returns the new EMA score.
+// If the inferer didn't miss any epochs this is a no-op, the EMA score is returned as is.
+func (k *Keeper) MayPenaliseInferer(
+	ctx sdk.Context,
+	topic types.Topic,
+	block types.BlockHeight,
+	emaScore types.Score,
+) (types.Score, error) {
+	return MayPenaliseActor(
+		ctx,
+		CountWorkerContiguousMissedEpochs,
+		func(topicId TopicId) (alloraMath.Dec, error) {
+			return k.GetTopicInitialInfererEmaScore(ctx, topicId)
+		},
+		func(topicId TopicId, score types.Score) error {
+			return k.SetInfererScoreEma(ctx, topicId, score.Address, score)
+		},
+		topic,
+		block,
+		emaScore,
+	)
+}
+
+// MayPenaliseForecaster penalises a forecaster for missing previous epochs. It saves and returns the new EMA score.
+// If the forecaster didn't miss any epochs this is a no-op, the EMA score is returned as is.
+func (k *Keeper) MayPenaliseForecaster(
+	ctx sdk.Context,
+	topic types.Topic,
+	block types.BlockHeight,
+	emaScore types.Score,
+) (types.Score, error) {
+	return MayPenaliseActor(
+		ctx,
+		CountWorkerContiguousMissedEpochs,
+		func(topicId TopicId) (alloraMath.Dec, error) {
+			return k.GetTopicInitialForecasterEmaScore(ctx, topicId)
+		},
+		func(topicId TopicId, score types.Score) error {
+			return k.SetForecasterScoreEma(ctx, topicId, score.Address, score)
+		},
+		topic,
+		block,
+		emaScore,
+	)
+}
+
+// MayPenaliseReputer penalises a reputer for missing previous epochs. It saves and returns the new EMA score.
+// If the reputer didn't miss any epochs this is a no-op, the EMA score is returned as is.
+func (k *Keeper) MayPenaliseReputer(
+	ctx sdk.Context,
+	topic types.Topic,
+	block types.BlockHeight,
+	emaScore types.Score,
+) (types.Score, error) {
+	return MayPenaliseActor(
+		ctx,
+		CountReputerContiguousMissedEpochs,
+		func(topicId TopicId) (alloraMath.Dec, error) {
+			return k.GetTopicInitialReputerEmaScore(ctx, topicId)
+		},
+		func(topicId TopicId, score types.Score) error {
+			return k.SetReputerScoreEma(ctx, topicId, score.Address, score)
+		},
+		topic,
+		block,
+		emaScore,
+	)
+}
+
+func MayPenaliseActor(
```

Comment:
Don't love the name but it's fine.

Great abstractions.

Reply:
Agree on both hehe

Reply:
Didn't find better but we can change it :)

Reply:
ApplyLivenessPenaltyToActor